assets/js/app.js

```diff
@@ -112,83 +118,91 @@ $(function() {
+
+          $row
+            .append( $('<td>').append($btn) )
+            .append($var);
+
+          $("#varstbody").append($row);
+
+          $('.btn-var-'+name).tooltip({
+            placement: 'right',
+            title: variable.type
+          });
+      });
     }
-    else if(!_.isEmpty(device_vars) && _.isEmpty(data.body.variables)){
-      $("#varstbody").html('<td colspan="2" class="centered"><i>None exposed by firmware</i></td>');
-    }
-    _.each(device_vars, function(value, key) {
-      if(!data.body.variables.hasOwnProperty(key)){
-          delete device_vars[key];
-      }
-    });
-    _.each(data.body.variables, function(value, key) {
-      if(!device_vars.hasOwnProperty(key)){
-        device_vars[key]="";
-        device_vartypes[key]=value;
-        $("#varstbody").append('<tr><td>'+key+'</td>'+
-                               '<td id="'+key+'">?</td></tr>');
-      }
-    });
 
-    $("#funcs").html('');
-    _.each(data.body.functions, function(item, idx) {
-      $("#funcs").append('<option id="'+item+'">'+item+'</option>');
-    });
+    if(_.isEmpty(data.body.functions)){
+      $("#funcs").html('<option>--</option>');
+    } else{
+      _.each(data.body.functions, function(item, idx) {
+        $("#funcs").append('<option>', {value: item, html: item});
+      });
+    }
   }
 
   function get_variables(){
-    console.log('get_variables()');
-    _.each(device_vars, function(value, key) {
-      particle.getVariable({deviceId: current_device, name: key, auth: access_token})
+    _.each(device_vars, function(variable, name) {
+      particle.getVariable({deviceId: current_device, name: name, auth: access_token})
         .then(update_variable)
     });
   }
 
   function update_variable(data){
-    console.log('update_variable(): ',data.body.name,'('+device_vartypes[data.body.name]+') old val:',device_vars[data.body.name],' new val:',data.body.result)
-    if(device_vars[data.body.name] != data.body.result){
-      device_vars[data.body.name]=data.body.result;
-      if(device_vartypes[data.body.name] == 'double'){
-        if(data.body.result == null){
-          $("[id='"+data.body.name+"']").html('NaN/Inf');
-        }
-        else{
-          $("[id='"+data.body.name+"']").html(data.body.result.toPrecision(6).toString());
-        }
-      }
-      else if(device_vartypes[data.body.name] == 'string'){
-        var str=data.body.result;
-        if(str.length>15){
-            str=str.substring(0,13)+'..';
-        }
-        $("[id='"+data.body.name+"']").html(str);
-      }
-      else{
-        $("[id='"+data.body.name+"']").html(data.body.result);
-      }
+    device_vars[data.body.name].value = data.body.result;
+    var localVar = device_vars[data.body.name];
+    var result = "";
+
+    if((localVar.type == 'double') && (data.body.result == null)){
+      result = 'NaN/Inf';
     }
+
+    if(localVar.type == 'bool')
+      result = data.body.result.toString();
```

Comment:
The [particle docs](https://docs.particle.io/reference/firmware/photon/#particle-variable-) claim to only support INT32, DOUBLE and STRING. However, based on [the OakCore sources](https://github.com/digistump/OakCore/blob/master/cores/oak/OakParticle/particle_core.cpp#L2119), it does look like BOOL is an undocumented extra.